### ISSUE TITLE
"Parameter hint" problem

### ISSUE THREAD

**Issue by lh123** [microsoft/vscode-cpptools#902]
- VsCode Version ： 1.14.1
- Extension Version：0.12.1
- Os : Win10 64Bit
- gcc : MinGW-W64(5.3.0)

1. Parameter hints can not be triggered when direct initialization is used。
2. Parameter hints can not be triggered again in parentheses

```c++
#include <iostream>
#include "lib/people.h"
using namespace std;
int main() {
    People p1("test",10);//Parameter hints can not be triggered when direct initialization is used。
    People *p2 = new People("test",10);//There are parameter hints
    People *p3 = new People("test",);//Parameter hints can not be triggered again in parentheses
    cin.get();
    return 0;
}
```

3. Automatic completion without context analysis, such as the prompt "cout" but `using namespace std` has been commented out

![image](https://user-images.githubusercontent.com/10697249/28402081-7da74a3a-6d50-11e7-9bfb-e6ff99f62db3.png)

Sorry my English is very bad

**Comment by sean-mcmanus** [comment]
As a workaround, you can get parameter hints a 2nd time by doing Ctrl+Shift+Space after the opening (, for the ("test", ) case. Parameter hints with direct initialization doesn't even work in VS 2017, but it works with the {} initialization syntax (we still don't). We don't yet have the semantic aware completion for non-prefixed words (only after -> . ::).

**Comment by lh123** [comment]
@sean-mcmanus Thank you for your answer
But I found that only re-enter the left parenthesis to re-pop up the parameter prompt, and <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>Space</kbd> in the brackets can not trigger the parameter prompt again.

#819 also has the same problem

**Another question**
When I #include <windows.h> in the file, Intellisense will produce many errors

- my code
```c++
#include <windows.h>
#include <iostream>

int main(){
    return 0;
}
```
![image](https://user-images.githubusercontent.com/10697249/28449405-5aee779e-6e12-11e7-9ea3-72dedbb4bf18.png)

- c_cpp_properties.json

```json
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "${workspaceRoot}",
                "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/include",
                "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/include-fixed",
                "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/../../../../x86_64-w64-mingw32/include",
                "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++",
                "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++/x86_64-w64-mingw32",
                "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++/backward"
            ],
            "defines": [
                "__GNUC__=5",
                "__cdecl=__attribute__((__cdecl__))"
            ],
            "intelliSenseMode": "clang-x64",
            "browse": {
                "path": [
                    "${workspaceRoot}",
                    "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/include",
                    "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/include-fixed",
                    "D:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.4.0/../../../../x86_64-w64-mingw32/include",
                    "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++",
                    "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++/x86_64-w64-mingw32",
                    "D:/mingw64/lib/gcc/../../x86_64-w64-mingw32/include/c++/backward"
                ],
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            }
        }
    ],
    "version": 2
}
```
How should I solve this problem?

**Comment by sean-mcmanus** [comment]
I don't think windows.h is intended to be compiled with the clang-x64 compiler. I assume it requires the msvc-x64 intelliSenseMode set. I thought mingw was meant for compiling Linux code and not Windows OS code.

**Comment by bobbrow** [comment]
I can take a look at this.  windows.h ships with MinGW and you can create Windows apps with it.

**Comment by lh123** [comment]
@sean-mcmanus Thank you for your answer, but I can not fully understand what you said.
"Windows.h" is the file that comes with MinGW-W64, and I can use the "g ++ main.cpp -o main" command to complete the compilation.
When I set intelliSenseMode to msvc-x64 mode, these errors disappear, but msvc-x64 does not work on MinGW.

**Comment by sean-mcmanus** [comment]
@lh123 Yeah, ignore what I said. Bob is more familiar with MinGW, so he'll investigate.

**Comment by lh123** [comment]
@bobbrow In the latest version, I found that adding `__stdcall=__attribute__((__stdcall__))` in the definition, IntelliSence can handle windows.h correctly.

**Comment by bobbrow** [comment]
I'm so sorry I dropped the ball on this.  We've seen a large increase in the number of issues reported per week as more people start using the extension and it's getting harder to keep track of them all.

**Comment by sean-mcmanus** [comment]
So should the MinGW docs be updated to include the __stdcall definition? https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/MinGW.md

**Comment by lh123** [comment]
@sean-mcmanus If this is correct, I think the MinGW docs should be updated

**Comment by bobbrow** [comment]
I was starting to look at this before I took a few days off.  MinGW-w64 seems to behave differently than my other MinGW installation (which defines __stdcall in the Windows headers without manual intervention by the c_cpp_properties.json file).

I'm still looking at this when I'm not getting pulled in to other issues, but I haven't settled on a final answer yet.  We may need to have some footnotes for the different distributions of MinGW.

**Comment by lh123** [comment]
@bobbrow I think you can use the `g++ -dM -E - /dev/null` command to automate the addition of some macro definitions.

**Comment by sean-mcmanus** [comment]
@lh123 Thanks for the info. Our pending update will try to get the default includePath from the compiler, but the defines might not be till a later update.

**Comment by xahon** [comment]
Parameter hint works only once when i open paren.
If i `esc` currently opened parameter hint infobox and press `ctrl+shift+space` nothing happens.

So i noticed that smart pointer constructor and their make don't show object constructor parameters.
`std::make_shared<int>(/* parameter hint will show shared_ptr contructor hints instead of int's */` - so may be it's right by design but i think it would be better if object's parameters will be shown

**Comment by bobbrow** [comment]
This should be fixed with version 0.14.3
